GUIDELINES.md

```diff
@@ -0,0 +1,153 @@
+
+
+## Design patterns
+
+Favour the following design patterns, in this order:
+
+1. Functional programming: using functions to do one thing well, not altering the original data.
+2. Modular code: using functions to eliminate duplication, improve readability and reduce indentation.
+3. Object-oriented programming: Generally avoid, unless you are customising an API (for example `DataFrame`) or defining your own API.
+
+If you are not, at least, adhering to a modular style then you have gone very wrong.
+You should implement unit tests for each of your functions, something which is generally more tricky for object-oriented programming.
+
+## Programming
+
+### Mandatory
+
+* Don't compare boolean values to `True` or `False`.
+* Favour `is not condition` over `not condition is`
+* Don't compare a value to `None` (`value == None`), always favour `value is None`
+
+### Encouraged
+
+* Favour `logging` over `print`
+* Favour using `GLOBAL_VARIABLES` near the top of your code rather than *any* hard-coded variables in your code. These can be more easily factored out into config, and at least puts many of your design decisions in one place.
+
+
+## Critical thinking
+
+* Surely this simple problem already has an elegant solution?
+* How many copies of the data am I making?
+* Where do my variables go out of scope?
+* How can I avoid creating a new variable at all costs? (think iterator, scope, `lru_cache`)
+
+## Naming conventions
+
+### Mandatory
+
+* Functions / methods: `function`, `my_function`
+* Variables / attributes: `variable`, `my_var`
+* Class: `Model`, `MyClass`
+* Module / file names / directory names: `module`, `file_name.py`, `dir_name`
+* Global* / constants: `A_GLOBAL_VARIABLE`
+
+\* Global at the module level, not the `global` level. Don't use `global`, ever.
+
+### Encouraged
+
+* Keep all names as short and descriptive as possible. Variable names such as `x` or `df` are highly discouraged unless they are genuinely representing abstract concepts.
+* Favour good naming conventions over helpful comments
+
+
+## Spaces and spacing
+
+### Encouraged
+
+* Use the absolute minimum number of indents of your code. You can achieve this by writing modular code, and inverting logic, for example:
+
+```python
+def something(args):
+	for item_collection in args:          # 1 tab
+		if item_collection.exists():      # 2 tabs
+			the_sum = 0                   # 3 tabs
+			for item in item_collection:
+				the_sum += item.value()   # 4 tabs
+			print(the_sum)
+```
+
+can become:
+
+```python
+def sum_values(item_collection):
+	the_sum = sum(item.value() for item in item_collection)
+	print(the_sum)
+
+def something(args):
+	for item_collection in args:          # 1 tab
+		if not item_collection.exists():  # 2 tabs
+			continue
+		sum_values(item_collection)
+```
+
```

Comment:
Or even

```python

from functools import reduce

reduce(lambda acc, item: acc + item.value(),
 filter(lambda item_collection: item_collection.exists(), args),
0
)
```

With toolz:
```python
from functools import reduce
import toolz.curried as t

t.pipe(args,
 t.filter(lambda item_collection: item_collection.exists()),
 lambda args: reduce(lambda acc, item: acc + item.value(), args,0)
)
```

Most loops are just re-implementations and combinations of `map`, `filter`, `reduce` but with bugs added in.

Reply:
At some point I'd like to do a FP workshop.

Reply:
**I actually misread this example so the above doesn't quite work**